src/main/java/frc/robot/subsystems/carriage/CarriageIOReal.java

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.subsystems.carriage;
+
+import com.ctre.phoenix6.BaseStatusSignal;
+import com.ctre.phoenix6.StatusSignal;
+import com.ctre.phoenix6.configs.TalonFXConfiguration;
+import com.ctre.phoenix6.controls.VoltageOut;
+import com.ctre.phoenix6.hardware.TalonFX;
+import com.ctre.phoenix6.signals.InvertedValue;
+import frc.robot.utils.components.ReversibleDigitalInput;
+
+/** Create a CarriageIO that uses a real TalonFX. */
+public class CarriageIOReal implements CarriageIO {
+  TalonFX motor = new TalonFX(0);
+
+  ReversibleDigitalInput beambreak = new ReversibleDigitalInput(0, false);
+
+  VoltageOut voltageOut = new VoltageOut(0.0).withEnableFOC(true);
+
+  StatusSignal<Double> velocity = motor.getVelocity();
+  StatusSignal<Double> voltage = motor.getMotorVoltage();
+  StatusSignal<Double> amperage = motor.getStatorCurrent();
+  StatusSignal<Double> temp = motor.getDeviceTemp();
+
+  public CarriageIOReal() {
+    var config = new TalonFXConfiguration();
+
+    config.MotorOutput.Inverted = InvertedValue.CounterClockwise_Positive;
+
+    motor.getConfigurator().apply(config);
+    BaseStatusSignal.setUpdateFrequencyForAll(50.0, velocity, voltage, amperage, temp);
+    motor.optimizeBusUtilization();
+  }
+
```

Comment:
Also worth noting that the patterns in both the constructor (setup signals properly) and the updateInputs implementation feel like they're going to be common enough that we should think about good ways to DRY them up. Java has relatively limited ways to do that and some are ugly (annotation generators and awkward reflection, yuck), but I'll have a think on it.

Reply:
for posterity there was a conversation on slack about this afterwards. my hope is to make a vscode extension for code gen rather than have something in the repository to generate code at compile time, since its easier to modify code generated by the editor than code generated at compile time